flow/project.py

```diff
@@ -3321,8 +3321,16 @@ def _execute_operation(self, operation, timeout=None, pretend=False):
                 operation,
                 operation.cmd,
             )
-            with self._run_with_hooks(operation):
-                subprocess.run(operation.cmd, shell=True, timeout=timeout, check=True)
+            try:
+                with self._run_with_hooks(operation):
+                    subprocess.run(
+                        operation.cmd, shell=True, timeout=timeout, check=True
+                    )
+            except subprocess.CalledProcessError as error:
+                raise UserOperationError(
+                    f"An exception was raised during operation {operation.name} "
+                    f"for job or aggregate with id {get_aggregate_id(operation._jobs)}."
+                ) from error
```

Comment:
```suggestion
            with self._run_with_hooks(operation):
                try:
                    subprocess.run(
                        operation.cmd, shell=True, timeout=timeout, check=True
                    )
                except subprocess.CalledProcessError as error:
                    raise UserOperationError(
                        f"An exception was raised during operation "
                        "{operation.name} for job or aggregate with id "
                        "{get_aggregate_id(operation._jobs)}."
                    ) from error
```

Reply:
This may not read the user-defined error messages, hence the tests would fail eventually. We may need to fix the behaviour of how we show errors to our users. I guess this relates to my comment here https://github.com/glotzerlab/signac-flow/issues/571#issuecomment-1008668958

A quick fix to this problem is simply to change the hooks tests but I guess this can be a good topic of discussion.

Reply:
That is a good point. Looking at the code it seems that the `on_fail` hook catches the error and re-raises it. Thanks for catching that @kidrahahjo!